src/process/ProcessManagerImpl.cpp

```diff
@@ -235,13 +233,18 @@ ProcessManagerImpl::tryProcessShutdown(std::shared_ptr<ProcessExitEvent> pe)
         auto runningIt = mProcesses.find(pid);
         if (runningIt != mProcesses.end())
         {
-            // Mark it as "ready to be killed"
+            // Mark it as "ready to be killed" if caller retries.
             mKillable.push_back(runningIt->second);
-            impl->cancel(ec);
-            auto res = cleanShutdown(*runningIt->second);
-            mProcesses.erase(pid);
-            gNumProcessesActive--;
-            return res;
+
+            // Send a SIGTERM / Ctrl-C
+            cleanShutdown(*runningIt->second);
+
+            // We initiated a clean shutdown here but the process isn't dead
+            // yet; caller needs to try again if they want to be sure.
+            //
+            // SIGCHLD and waitpid will be handled up above in
+            // handleSignalWait / handleProcessTermination.
+            return false;
```

Comment:
I think this commit broke clean shutdown for `RunCommand`?

before: it would return `true` not because the process was terminated, but because the call `kill(pid, SIGINT)` succeeds. I am not sure when force kill was called in this particular situation (I think when the main process really stops)... but most commands that we call would not need to be force killed.

now: it returns false, causing `onAbort` to retry in the next `crank` call, which is likely to happen very fast, and when it gets called again, it does a force kill.

This can have bad effects as "force killing" scripts (like what can be done with history archives get/put) that spawn subprocesses would not be able to cleanup properly. This increases the chances of corruptions in those scripts/sub-processes.

Reply:
I agree this is the effect of it but I'm not sure what the right thing to do is instead. Maybe have some sort of shutdown/grace period? We should force-kill _eventually_ if TERM isn't effective.

Reply:
There are probably a couple issues:
* we should transition from `ABORTING` to `ABORTED` only when we have confirmation that the process is actually dead
* we already "kinda have" a grace period between `ApplicationImpl::gracefulStop()` and `shutdownMainIOContext` (and then the actual "application/process manager destructor" that will force kill everything)
   * but I don't know if having some work in the `ABORTING` state actually stops the main loop from exiting

The cleanest thing to do might be to add a grace period (shorter than the one used in the application) to `RunCommand` - but I think that for that to work, we need a new work state `ABORTING_WAITING` (that avoids busy loops).

Reply:
decided to back off the "guaranteed kill" approach: maintaining the old/existing API contract which is "report true when we successfully _signalled_ the process to exit (and cancelled the client's completion event), let ProcessManager itself mop up the actual exit".

I don't love this contract, but it works and it seems like replacing it just shifts the possible failure modes around without really fixing them. Best to leave as-is.